nixos/modules/system/activation/switch-to-configuration.pl

```diff
@@ -564,86 +619,58 @@ sub handle_modified_unit { ## no critic(Subroutines::ProhibitManyArgs, Subroutin
-}
-foreach (@upcomingNspawnUnits) {
-    my $unit = basename($_);
-    $unit =~ s/\.nspawn//;
-    my $unitName = "systemd-nspawn\@$unit.service";
-    my $orig = $_;
-    $orig =~ s/^$out//;
-    if ($orig ~~ @currentNspawnUnits) {
-        if (fingerprintUnit($_) ne fingerprintUnit($orig)) {
-            my ($eq, $strategy) = compareNspawnUnits($orig, $_);
-            if ($strategy ne "none") {
-                if ($strategy ne "restart" and ($eq == 0 or $strategy eq "reload")) {
-                    if (isContainerRunning($unit) == 1) {
-                        $units_to_reload{$unitName} = 1;
-                    }
-                } elsif ($eq == 1) {
-                    $units_to_restart{$unitName} = 1;
-                }
+# Handle nspawn unit changes
+my @current_nspawn_units = glob("/etc/systemd/nspawn/*.nspawn");
+my @new_nspawn_units = glob("$out/etc/systemd/nspawn/*.nspawn");
+my %current_units_cmp = map { $_ => 1 } @current_nspawn_units;
+foreach my $new_unit_file (@new_nspawn_units) {
+    my $container_name = basename($new_unit_file);
+    $container_name =~ s/\.nspawn//;
+    my $unit_name = "systemd-nspawn\@$container_name.service";
+
+    my $cur_unit_file = $new_unit_file;
+    $cur_unit_file =~ s/^$out//;
+    if (exists($current_units_cmp{$cur_unit_file})) {
+        my %new_unit_info = parse_unit($new_unit_file);
+        my $strategy = $new_unit_info{"Exec"}{"X-ActivationStrategy"}[0] // "dynamic";
+
+        # Skip comparison logic/restart check if ActivationStrategy is "none"
+        next if $strategy eq "none";
+
+        my %cur_unit_info = parse_unit($cur_unit_file);
+        my $changed = compare_nspawn_units(\%cur_unit_info, \%new_unit_info);
+
+=pod Truth table for restarts
+|Strategy|Changed|Reload|Restart|
+|--------|-------|------|-------|
+|Dynamic |0      |Y     |-      |
+|Dynamic |1      |-     |Y      |
+|Restart |0      |-     |-      |
+|Restart |1      |-     |Y      |
+|Reload  |0      |Y     |-      |
+|Reload  |1      |Y     |-      |
+=cut
+        if ($strategy ne "restart" and ($changed == 0 or $strategy eq "reload")) {
+            if (is_container_running($container_name) == 1) {
+                $units_to_reload{$unit_name} = 1;
             }
+        } elsif ($changed == 1) {
+            $units_to_restart{$unit_name} = 1;
         }
     } else {
-        $units_to_start{$unitName} = 1;
+        # Start the unit if it didn't exist before
+        $units_to_start{$unit_name} = 1;
     }
 }
 
-foreach (@currentNspawnUnits) {
-    unless (-f "$out$_") {
-        my $unit = basename($_);
-        $unit =~ s/\.nspawn//;
-        my $unitName = "systemd-nspawn\@$unit.service";
-        $units_to_stop{$unitName} = 1;
+# Stop all now removed nspawn containers
+foreach my $cur_unit_file (@current_nspawn_units) {
+    my %cur_unit_info = parse_unit($cur_unit_file);
+    unless (-f "$out$cur_unit_file" || $cur_unit_info{"Exec"}{'X-Imperative'}[0] == "1") {
```

Comment:
container-activation test fails here with `Use of uninitialized value in numeric eq (==) at /nix/store/bbx8m9sbwj0i9ivkk63dn85b3bw6mxbc-nixos-system-configchange-23.05pre-git/bin/switch-to-configuration line 669.`. Trying to fix.